domain/src/test/java/org/oppia/domain/oppialogger/exceptions/ExceptionsControllerTest.kt

```diff
@@ -0,0 +1,350 @@
+  }
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testController_logException_switchToNoNetwork_logException_checkManagement() {
+    val exceptionThrown = Exception("TEST", Throwable())
+    exceptionsController.logNonFatalException(exceptionThrown, TEST_TIMESTAMP)
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+    exceptionsController.logFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val cachedExceptions = exceptionsController.getExceptionLogs()
+    cachedExceptions.observeForever(mockOppiaExceptionLogsObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockOppiaExceptionLogsObserver, atLeastOnce())
+      .onChanged(oppiaExceptionLogsResultCaptor.capture())
+    val exceptionFromRemoteService = fakeExceptionLogger.getMostRecentException()
+    val exceptionFromCacheStorage =
+      oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(0)
+    val exception = exceptionLogToException(exceptionFromCacheStorage)
+
+    assertThat(exceptionFromRemoteService).isEqualTo(exceptionThrown)
+    assertThat(exception.message).isEqualTo(exceptionThrown.message)
+    assertThat(exception.stackTrace).isEqualTo(exceptionThrown.stackTrace)
+    assertThat(exception.cause?.message).isEqualTo(exceptionThrown.cause?.message)
+    assertThat(exception.cause?.stackTrace).isEqualTo(exceptionThrown.cause?.stackTrace)
+    assertThat(exceptionFromCacheStorage.exceptionType).isEqualTo(ExceptionType.FATAL)
+  }
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testController_logNonFatalException_logFatalException_withNoNetwork_checkOrderinCache() {
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+    val exceptionThrown = Exception("TEST", Throwable())
+    exceptionsController.logNonFatalException(exceptionThrown, TEST_TIMESTAMP)
+    exceptionsController.logFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val cachedExceptions = exceptionsController.getExceptionLogs()
+    cachedExceptions.observeForever(mockOppiaExceptionLogsObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockOppiaExceptionLogsObserver, atLeastOnce())
+      .onChanged(oppiaExceptionLogsResultCaptor.capture())
+    val exceptionOne = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(0)
+    val exceptionTwo = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(1)
+    assertThat(exceptionOne.exceptionType).isEqualTo(ExceptionType.NON_FATAL)
+    assertThat(exceptionTwo.exceptionType).isEqualTo(ExceptionType.FATAL)
+  }
+
+  private fun setUpTestApplicationComponent() {
+    DaggerExceptionsControllerTest_TestApplicationComponent.builder()
+      .setApplication(ApplicationProvider.getApplicationContext())
+      .build()
+      .inject(this)
+  }
+
+  private fun exceptionLogToException(exceptionLog: ExceptionLog): Exception {
```

Comment:
This & below: these are a bit too much logic for a test. Could we maybe have a few assertion helpers that verify a few aspects of the exception rather than trying to do an exact comparison between the two? Or, factor out the prod conversion code into a helper that both the controller & its test can use, and have a few dedicated tests for that helper utility to verify it's correct.

The latter is probably the cleaner solution, but it's up to you which approach you want to go with.

Reply:
Working on it.

Reply:
Done.